Pecos hybrid channel testing

clarkpede opened the pull request:
This is PR 3 of a series of PRs intended to get the pecos-dev branch back in line with the "best" and most stable code we are currently using.

trevilo committed:
merge from stash.  Using this version for channel verification testing.

trevilo committed:
add factor inv(L)

trevilo committed:
write vtk with more precision

trevilo committed:
update conserved averaged prior to calculating k and production

trevilo committed:
undo shift in TG forcing for channel

trevilo committed:
Track mean of forcing so we can subtract it off.

trevilo committed:
output mean of forcing

trevilo committed:
switch back to Sigfried's aspect ratio blend in M43

trevilo committed:
remove trace of M43 stress in resolution adequacy calc

trevilo committed:
make sure IMEX schems are treated as implicit

trevilo committed:
lower minimum in Roe low dissipation schemes

trevilo committed:
make SMR91 time marching default to 3 substeps

trevilo committed:
Fix pressure (by including alpha), and change production time scale in averaging to 4x

trevilo committed:
Modify forcing to account for y/z swap in my channel grid, and remove min against wall distance in length scale

trevilo committed:
Back to forward Euler integration for averages

trevilo committed:
Playing with damping on M43

trevilo committed:
Switch to a*(2-a) scaling on anisotropic part of tauSGRS

trevilo committed:
Increase min upwinding to 0.01, for stability

trevilo committed:
Merge from pecos-dev (fixes to rho dependence in v2-f in particular)

trevilo committed:
In order to get recently added input file options,
merge branch 'pecos-dev' into pecos-hybrid-channel-testing

 Conflicts:
	SU2_CFD/src/numerics_direct_turbulent_v2f.cpp
	SU2_CFD/src/solver_direct_turbulent_v2f.cpp
	config_template.cfg

trevilo committed:
hybrid channel is stable with these mods... backward euler for averaging is important?

trevilo committed:
Merge branch 'pecos-investigate-hybrid-channel-stability' into pecos-hybrid-channel-testing.

After fixes that went into v2-f and hybrid model, the Retau=5200
channel case was not stable in hybrid mode.  Pulling changes from
branch where I investigated that issue.  On that branch, stability was
much improved using backward Euler to advance the average state.  Will
try in combination with other fixes now.

trevilo committed:
first pass at M43 Jacobians

clarkpede committed:
Merge branch 'pecos-dev' into pecos-hybrid-channel-testing

clarkpede committed:
Add config option for linear solver verbosity

clarkpede commented on SU2_CFD/include/hybrid_RANS_LES_forcing.inl:
```diff
@@ -91,9 +92,22 @@ inline void CHybridForcingTG0::SetTGField(
+  // CHANNEL HACK
+  h[0] = A * cos(a[0]*x[0]) * sin(a[1]*x[1]) * sin(a[2]*(x[2]-1.0));
+  h[1] = B * sin(a[0]*x[0]) * cos(a[1]*x[1]) * sin(a[2]*(x[2]-1.0));
+  h[2] = C * sin(a[0]*x[0]) * sin(a[1]*x[1]) * cos(a[2]*(x[2]-1.0));
```
@trevilo Why is this shift (`x[2] - 1`) necessary?

trevilo replied:
I don't think it is necessary.  I originally shifted z (my wall-normal coordinate) b/c Sigfried had shifted y (his wall-normal coordinate) in his first testing in CDP.  He did this for symmetry purposes, but later stopped b/c it didn't seem to matter much (or that is my understanding at least).  In any case, I don't think this shift combined with the form here actually gives the right symmetry.  It is symmetric in the wall-normal component when you would want that to be anti-symmetric, right?  This is true in CDP as well.  So, bottom line is that I think this can be safely eliminated.

clarkpede replied:
Alright.  I'll compare a short run both with and without the shift.

clarkpede replied:
The results are essentially the same:
![Figure_6](https://user-images.githubusercontent.com/13340225/64719527-0a042880-d48e-11e9-8dd8-f63b4f18483c.png)

trevilo replied:
Looks good.

clarkpede commented on SU2_CFD/src/fluctuating_stress.cpp:
```diff
@@ -111,41 +111,90 @@ void CM43Model::CalculateEddyViscosity(const CGeometry* geometry,
+      //C_M0*C_M * pow(dissipation, 1.0/3) * M43[iDim][jDim];
+	//C_M0*C_M * pow(dissipation, 1.0/3) * M43[iDim][jDim];
+	//density * C_M0*C_M * pow(dissipation, 1.0/3) * M43[iDim][jDim];
+	//density * damping*C_M0*C_M * pow(dissipation, 1.0/3) * M43[iDim][jDim];
```
@trevilo We calculate damping here, but we don't use it? What's the status on the damping?

trevilo replied:
Yeah, the damping or wall-blending of the SGET term is something I messed around with, but without ever coming to a definitive conclusion for how it should be handled.  You can see evidence of this in the code just below, where as similar blending is applied as what is in CDP (although it blends to 0 rather than the RANS eddy viscosity b/c I had stability problems when blending to the RANS eddy viscosity).

In any case, Sigfried did not need this blending for the channel in CDP.  He added it later for very high aspect ratio cells in the WMH.  Empirically, I observed that we did need something (for stability) in SU2, and the stability of the code was somewhat sensitive to how exactly it was done.  However, those observations are mostly from when the SGET term was treated fully explicitly.  So, it should be less sensitive now I would hope.

clarkpede replied:
Alright.  I'm moving this to a config option, with three choices:

1. No high-AR damping
2. Blend to zero for high-AR
3. Blend to RANS for high-AR

clarkpede commented on SU2_CFD/src/hybrid_RANS_LES_model.cpp:
```diff
@@ -351,21 +352,22 @@ void CHybrid_Mediator::SetupResolvedFlowSolver(const CGeometry* geometry,
+    //   for (unsigned short jDim = 0; jDim < nDim; jDim++) {
+    //      aniso_eddy_viscosity[iDim][jDim] *= factor;
+    //   }
+    // }
```
@trevilo Is there a reason this was removed?  Was it also removed in CDP?

trevilo replied:
No, it is still in CDP.  I commented it out for the channel runs for two reasons: 1) it is hacky so I just wanted to see if we really needed it (at least for the channel), and 2) for stability again.  Now, as before, the stability issues were observed prior to any implicitness on the SGET term.  So, I was trying to keep mu^{SGET} from getting big.  That shouldn't really be a concern anymore, but I have not tested with this back on.

clarkpede commented on SU2_CFD/src/hybrid_RANS_LES_model.cpp:
```diff
@@ -534,6 +536,15 @@ void CHybrid_Mediator::ComputeInvLengthTensor(CVariable* flow_vars,
+  for (iDim = 0; iDim < nDim; iDim++) {
+    tauSGET[iDim][iDim] -= traceTauSGET/3;
+  }
+
```
@trevilo Shouldn't we subtract out the trace of tauSGET where it's defined in the CAvgGrad_Hybrid as well?

In the long term, we need to find a way to eliminate the duplication between numerics and the InvLengthTensor calculation.  I've got an idea or two, but I haven't had time to work on them.

trevilo replied:
Oops, I think I missed this one the first time through.

Anyway... yeah, it should be consistent.  I don't really know that trace should be subtracted out in either spot.  In the model, the trace of tauSGET contributes to the modeled stress in the same way as (2/3)*alpha*k, so it at least ought to average to zero to avoid contaminating k.  The easiest way to achieve that is simply to have the trace be zero.  I'm pretty sure it was there at one point in CAvgGrad_Hybrid, b/c I remember commenting it out while investigating stability problems.

Bottom line: I'm fine with including it in CAvgGrad_Hybrid too, assuming stability is ok.

clarkpede commented on SU2_CFD/src/hybrid_RANS_LES_model.cpp:
```diff
@@ -584,10 +599,12 @@ void CHybrid_Mediator::ComputeInvLengthTensor(CVariable* flow_vars,
         }
         // rho*k contribtuion
-        Pij[iDim][jDim] -= 2.0*alpha*rho*ktot*(Sd[iDim][jDim]+Om[iDim][jDim])/3.0;
+	// should this part be just alpha scaling?
```
Yes, yes it should.

trevilo replied:
I agree, which is why the code is that way.  But, just food for thought, Sigfried says he gets better results on the periodic hill when the whole stress is scaled by a*(2-a), not just the anisotropic part. Given the definition of alpha, scaling by a*(2-a) on the isotropic part is clearly an inconsistency.  So... the fact that that works better must mean it is covering for some other issue.

clarkpede committed:
Add config option to set the minimum upwinding

clarkpede commented on SU2_CFD/src/solver_direct_mean.cpp:
```diff
@@ -17702,24 +17715,36 @@ void CNSSolver::Source_Residual(CGeometry *geometry, CSolver **solver_container,
+	//Residual[nDim+1] += -Volume * U[iDim+1] * (Force[iDim] - MeanForce[iDim]);
+	Residual[nDim+1] += 0.0; // no source in energy?
+	//Residual[nDim+1] += -Volume * (U[iDim+1]/U[0]) * (Force[iDim] - MeanForce[iDim]);
+      }
```
@trevilo What's going on here?

trevilo replied:
Ha, just me being dumb.  I originally had the forcing entering the total energy equation, since that is what you'd do for a "normal" forcing term (i.e., a source in the momentum equation).  But, the forcing here is a transfer from unresolved to resolved.  Therefore, there isn't any forcing contribution to the total energy equation.

clarkpede committed:
Fix bug in RK IMEX schemes.

Previously, if no parameters were specified, the default RK
parameters were set in `Config::SetOutput`.  But `Config::SetOutput` is
only called by the rank 0 MPI task.  So the rest of the nodes were left
with uninitialized RK parameters.

This commit moves the default initialization of the RK schemes to a more
appropriate place.

clarkpede committed:
Move the fluctuating-term damping to a cfg option

clarkpede committed:
Update and remove comments for clarity

clarkpede committed:
Update config template with new options

clarkpede committed:
Remove error checking for SMR91 coeff.

clarkpede commented on SU2_CFD/src/solver_structure.cpp:
```diff
@@ -3025,7 +3025,8 @@ void CSolver::UpdateAverage(const su2double weight, const unsigned long iPoint,
   for (unsigned short iVar = 0; iVar < nVar; iVar++) {
-    const su2double new_average = (current[iVar] - average[iVar])*weight + average[iVar];
+    // BE for everything
+    const su2double new_average = (current[iVar]*weight + average[iVar])/(1.0 + weight);
```
If we use BE here, we should use BE everywhere (including for our hybrid variables like production).  Right now CNSSolver::UpdateAverage still uses FE.  I can add a fix to change the weight to be the "weight" for BE, which will make everything consistent again.

I don't know if this will change things much, but we should at least be consistent.

trevilo replied:
I agree: they should all be the same.  At one point early on when I was having stability problems with the channel I changed them all to BE, just to see if it helped.  It didn't, and I don't think it should affect anything much one way or the other, but I guess I forgot to switch these back.  Anyway, we can go back to FE or add an input option.  Whatever you prefer.

clarkpede commented:
So there is a slight change after ~6 flow-throughs (3000 timesteps) with my recent "clean-up" commits.  I tracked the change down to commit f074f3325c0834b8c84b77f963d168e7cc2d016d.  I changed the boundary conditions in v2f to fix what I think was a bug.

I'll do a longer run to see how much things differ over 7000 timesteps (~13 flow-throughs).  If we're still happy with the results, that will be a second "baseline" against which I'll compare future changes.

trevilo commented:
> So there is a slight change after ~6 flow-throughs (3000 timesteps) with my recent "clean-up" commits. I tracked the change down to commit [f074f33](https://github.com/pecos-hybrid/SU2/commit/f074f3325c0834b8c84b77f963d168e7cc2d016d). I changed the boundary conditions in v2f to fix what I think was a bug.

Regarding that BC fix, I agree there was a bug in using the instantaneous flow rather than the mean.  I have a similar mod on the axibump testing branch.  I used iPoint rather than jPoint (on the wall rather than 1 off) for the density and viscosity.  I'm not sure off the top of my head which it should actually be, so... just something to check.

clarkpede commented:
> I used iPoint rather than jPoint (on the wall rather than 1 off) for the density and viscosity. I'm not sure off the top of my head which it should actually be, so... just something to check.

I used the first off-node value because that's what's done in SST.  I did check it just now, though.  [Kalitzin](https://arc.aiaa.org/doi/pdf/10.2514/6.1999-3780), explicitly stated that he used the first point off the wall for his values of k and y that appear in the boundary conditions.  Unfortunately, that paper was incompressible, so viscosity and density were uniform.  The only transonic v2-f paper I know doesn't specify which point was used.  The NASA TMR specifies the first off-wall node.

The first off-wall node makes sense to me.  If I understand the BC correctly, we're trying to indirectly enforce a derivative boundary condition.  Therefore, we need the values at the first off-wall node and the values at the wall.  But since k_w = 0, the wall terms disappear, and we're only left with the first node terms.  Please correct me if I'm wrong.

clarkpede commented:
So here's the difference after 7000 timesteps.  The picture is very similar for 6000 and 6500 timesteps.  These are all calculated with planar averages, which means "u" is the planar average of the instantaneous velocity.

![Long_time_PR_v2f](https://user-images.githubusercontent.com/13340225/64550956-a34e0600-d2f9-11e9-817b-a9005fec0c14.png)

`PR_1` is commit f074f33 and baseline is the `channel_lm` tag.  They both look "right".  Having a higher alpha makes sense to me, since we're damping fluctuations by switching from instantaneous values to average values in the boundary conditions.

trevilo commented:
> I used the first off-node value because that's what's done in SST.

Ok, this seems like a good enough reason.  I think it actually doesn't really matter which we use (see below).

> The first off-wall node makes sense to me. If I understand the BC correctly, we're trying to indirectly enforce a derivative boundary condition. 

Yes, we're trying to get the right behavior of k, specifically k=0, which is set directly, and dk/dy=0, which is consistent with setting the right value of epsilon, namely \varepsilon(0) = \lim_{y \rightarrow 0} 2 \nu k / y^2.  

> Therefore, we need the values at the first off-wall node and the values at the wall. But since k_w = 0, the wall terms disappear, and we're only left with the first node terms. Please correct me if I'm wrong.

I don't entirely follow you here.  We need to approximate the limit above.  Since k and y are zero at the wall, you can't just evaluate it, and so we approximate it with data from the first point off the wall.  In incompressible solvers you get k and y at the first node off the wall, and since \nu is constant there is no ambiguity.  When \nu varies we have to make a choice, either evaluate the whole expression at the first node off the wall, or take \nu at the wall and k and y from the first node off.  Both are consistent approximations that will approach the true limit as the first point gets closer to the wall, so it isn't immediately obvious to me that we should prefer one over the other.

I kind of prefer to use \nu at the wall, just because there isn't any reason to approximate that part of the limit expression, but... that isn't really a strong reason, and I doubt it matters much either way.  So, I'm happy to leave it as you have it.

trevilo commented:
> `PR_1` is commit [f074f33](https://github.com/pecos-hybrid/SU2/commit/f074f3325c0834b8c84b77f963d168e7cc2d016d) and baseline is the `channel_lm` tag. They both look "right". Having a higher alpha makes sense to me, since we're damping fluctuations by switching from instantaneous values to average values in the boundary conditions.

Looks good.  I'm a tiny bit surprised it affects \alpha this much, but... fine.  I don't see any reason not to think it is right.

clarkpede committed:
Remove "-1" channel hack

This hack was originally designed to make the forcing completely
symmetric about the centerline.  In practice, this is not necessary.
Performance was compared for several flow-throughs, and results were
similar with and without the hack.

clarkpede committed:
Apply BE temporal discretization to all averages

Previous, backward Euler was used for the state variables but was not
applied to any of the extra variables (e.g. resolution adequacy). This
commit changes the weighting such that all averages are advanced using
a backward Euler scheme.

clarkpede committed:
Update tauSGET to remove trace.

clarkpede committed:
Update forcing test values to match new forcing.

In previous commits, the A/B/C constants of the Taylor green vortex and
the "-1" shift were both changed.  This changed the expected results of
the unit tests.  Unfortunately, this does mean that the code no longer
matches the previous unit test results extracted from CDP.

trevilo reviewed (commented):
@clarkpede do you have channel results with this mod (i.e., 5779003) or is that the case sitting in the queue?

trevilo commented on SU2_CFD/src/hybrid_RANS_LES_model.cpp:
```diff
@@ -561,6 +557,11 @@ void CHybrid_Mediator::ComputeInvLengthTensor(CVariable* flow_vars,
+   * make the turbulent kinetic energy inconsistent, so it is avoided here.
+   * But that indicates there's some other issue. ---*/
+
   su2double alpha_fac = alpha*(2.0 - alpha);
```
An update on this.  Sigfried indicates it isn't actually true.  I guess that conclusion (about scaling the whole stress by a*(2-a)) was based on looking at just one snapshot.  Anyway, now he says, "After many more runs, I was wrong about the scaling of the trace used in rM.  The best results come from alpha scaling on the isotropic part".

clarkpede committed:
Revert "Update tauSGET to remove trace."

This reverts commit 5779003e7c206a89b92f5f06634d8ff119b8fd15.
The tauSGET trace seems to cause stability issues.  This
will need further investigation.   The model (as currently
implemented) is inconsistent.

clarkpede commented:
Alright.  After reverting the tauSGET trace, we're back to a relatively normal run.  It looks like changing everything to BE did change the behavior a little, but we're still in the same ballpark.  Here's some comparisons between the original state (the `channel_lm` tag) and the current state:
![hybrid_26000](https://user-images.githubusercontent.com/13340225/65076177-10971200-d95e-11e9-978a-e2c7a40ebce7.png)
![hybrid_27000](https://user-images.githubusercontent.com/13340225/65076178-112fa880-d95e-11e9-8274-048c570cdfa3.png)

clarkpede committed:
Change resolution adequacy options

Remove the deprecated options.  Add a new option to use the full
velocity gradient tensor (including the divergence of velocity term).

trevilo commented on SU2_CFD/src/hybrid_RANS_LES_model.cpp:
```diff
@@ -580,18 +497,21 @@ void CHybrid_Mediator::ComputeInvLengthTensor(CVariable* flow_vars,
+          // rotation contribution
+          Pij[iDim][jDim] += 2.0*alpha_fac*eddy_viscosity*Sd_avg[iDim][kDim]*Om[kDim][jDim];
+          // Isotropic contribution (from delta_ij div_vel)
+          Pij[iDim][jDim] += 2.0*alpha_fac*eddy_viscosity*Sd_avg[iDim][jDim]*div_avg_vel/3.0;
```
Shouldn't this line use the divergence of the resolved velocity, not the average?

clarkpede replied:
You're correct.  I'll fix that.

trevilo commented on SU2_CFD/src/hybrid_RANS_LES_model.cpp:
```diff
@@ -580,18 +497,21 @@ void CHybrid_Mediator::ComputeInvLengthTensor(CVariable* flow_vars,
         }
-        // TODO: Add rho*k contribution
+        // rho*k contribtuion
+	Pij[iDim][jDim] -= 2.0*alpha*rho*ktot*(Sd[iDim][jDim]+Om[iDim][jDim])/3.0;
```
For RDELTA_INDICATOR_FULLP_WITH_DIV, isn't there a div(u) contribution to this term?

clarkpede replied:
Good catch.  I'll fix that.

clarkpede committed:
Fix bugs in resolution adequacy
